python-squid/src/main/java/org/sonar/python/metrics/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,232 @@
+import org.sonar.python.api.PythonGrammar;
+import org.sonar.python.api.PythonKeyword;
+import org.sonar.python.api.PythonPunctuator;
+
+public class CognitiveComplexityVisitor extends PythonVisitor {
+
+  private int complexity = 0;
+  private Deque<NestingLevel> nestingLevelStack = new LinkedList<>();
+
+  @Nullable
+  private final SecondaryLocationConsumer secondaryLocationConsumer;
+
+  public interface SecondaryLocationConsumer {
+    void consume(AstNode node, String message);
+  }
+
+  CognitiveComplexityVisitor(@Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    this.secondaryLocationConsumer = secondaryLocationConsumer;
+    nestingLevelStack.push(new NestingLevel());
+  }
+
+  public static int complexity(AstNode node, @Nullable SecondaryLocationConsumer secondaryLocationConsumer) {
+    CognitiveComplexityVisitor visitor = new CognitiveComplexityVisitor(secondaryLocationConsumer);
+    visitor.scanNode(node);
+    return visitor.complexity;
+  }
+
+  public int getComplexity() {
+    return complexity;
+  }
+
+  @Override
+  public Set<AstNodeType> subscribedKinds() {
+    return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+      PythonGrammar.IF_STMT,
+      PythonKeyword.ELIF,
+      PythonKeyword.ELSE,
+
+      PythonGrammar.WHILE_STMT,
+      PythonGrammar.FOR_STMT,
+      PythonGrammar.EXCEPT_CLAUSE,
+
+      PythonGrammar.AND_TEST,
+      PythonGrammar.OR_TEST,
+
+      PythonGrammar.TEST,
+
+      PythonGrammar.FUNCDEF,
+      PythonGrammar.CLASSDEF,
+      PythonGrammar.SUITE)));
+  }
+
+  @Override
+  public void visitNode(AstNode astNode) {
+    if (astNode.is(PythonGrammar.FUNCDEF, PythonGrammar.CLASSDEF)) {
+      nestingLevelStack.push(new NestingLevel(nestingLevelStack.peek(), astNode));
+    } else if (astNode.is(PythonGrammar.SUITE)) {
+      if (isSuiteIncrementsNestingLevel(astNode)) {
+        nestingLevelStack.peek().increment();
+      }
+    } else if (astNode.is(PythonGrammar.IF_STMT, PythonGrammar.WHILE_STMT, PythonGrammar.FOR_STMT, PythonGrammar.EXCEPT_CLAUSE)) {
+      incrementWithNesting(astNode.getFirstChild());
+    } else if (astNode.is(PythonKeyword.ELIF) || (astNode.is(PythonKeyword.ELSE) && astNode.getNextSibling().is(PythonPunctuator.COLON))) {
+      incrementWithoutNesting(astNode);
+    } else if (astNode.is(PythonGrammar.AND_TEST, PythonGrammar.OR_TEST)) {
+      incrementWithoutNesting(astNode.getFirstChild(PythonKeyword.AND, PythonKeyword.OR));
+    } else if (astNode.is(PythonGrammar.TEST) && astNode.hasDirectChildren(PythonKeyword.IF)) {
+      // conditional expression
+      incrementWithNesting(astNode.getFirstChild(PythonKeyword.IF));
+      nestingLevelStack.peek().increment();
+    }
+  }
+
+  @Override
+  public void leaveNode(AstNode astNode) {
+    if (astNode.is(PythonGrammar.FUNCDEF, PythonGrammar.CLASSDEF)) {
+      nestingLevelStack.pop();
+    } else if (astNode.is(PythonGrammar.SUITE)) {
+      if (isSuiteIncrementsNestingLevel(astNode)) {
```

Comment:
Same here, this inner if-condition could be merged with its parent, as well as the next condition since body are the same